### PR TITLE
introduced strong types

### PR DESCRIPTION
**Pull request by evaleev**
typical use is to distinguish creator/annihilator, particle/hole quantities, etc.

this is a preliminary implementation, sufficient to drive discussion in the Working Group

### CODE REVIEW

File: SeQuant/domain/mbpt/models/cc.cpp [hunk 1]
```diff
@@ -292,36 +292,34 @@ std::vector<ExprPtr> CC::eom_r(size_t nann, size_t ncre) {
...
 
-  using boost::numeric_cast;
   // start from the highest excitation order, go down to the lowest possible
-  for (auto na = numeric_cast<std::int64_t>(nann),
-            nc = numeric_cast<std::int64_t>(ncre);
-       na > 0 || nc > 0; --na, --nc) {
+  for (std::int64_t ra = na, rc = nc; ra > 0 || rc > 0; --ra, --rc) {
     // project with <ncre, nann| (i.e., multiply P(ncre, nann)) and compute VEV
-    result.at(idx) = vac_av(P(nc, na) * hbar_R, op_connect);
+    result.at(idx) = vac_av(P(nₚ(ra), nₕ(rc)) * hbar_R, op_connect);
```
Comment (evaleev) [inline SeQuant/domain/mbpt/models/cc.cpp:302]: @ajay-mk @bimalgaudel @cmasteran1 @Krzmbrzl comments welcome on such use of unicode (which avoids overly-short `np` and `nh`)

File: tests/unit/test_mbpt.cpp [hunk 2]
```diff
@@ -326,20 +326,20 @@ TEST_CASE("NBodyOp", "[mbpt]") {
             L"{{{\\frac{1}{2}}}{A^{{i_1}}_{{a_1}{a_2}}}{\\tilde{a}^{{a_1}{a_2}}"
             L"_{\\textvisiblespace\\,{i_1}}}}");
 
-    auto P_0_1 = P(0, 1)->as<op_t>();
+    auto P_0_1 = P(nₚ(1), nₕ(0))->as<op_t>();
     //    std::wcout << "P_0_1: " << to_latex(simplify(P_1_0.tensor_form()))
     //               << std::endl;
     REQUIRE(to_latex(simplify(P_0_1.tensor_form())) ==
             L"{{A^{}_{{i_1}}}{\\tilde{a}^{{i_1}}}}");
```
Comment (evaleev) [inline tests/unit/test_mbpt.cpp:333]: this clearly does not match `P_0_1` specification, since `nₕ` is 0

Comment (ajay-mk) [reply tests/unit/test_mbpt.cpp:333]: The same for the next two tests, `P_2_1` and `P_2_3` are also not correct

Comment (evaleev) [reply tests/unit/test_mbpt.cpp:333]: fixed in https://github.com/ValeevGroup/SeQuant/pull/228/commits/4d9fa6c118c9e362a8d6256c1d48f7423d70eeff ... this is probably why EOM-{IP,EA}-CC did not work

Comment (ajay-mk) [reply tests/unit/test_mbpt.cpp:333]: I think I manually checked everything when I initially wrote the tests, but it probably got mixed up later. I will be more careful next time. On that note, EOM-CCSD-R tests have been verified numerically, but not EOM-CCSD-L. We can remove it if not needed.

**Comment by evaleev** [comment]
@ajay-mk ready for review

**Review by ajay-mk — APPROVED** [review]
Looks good to me. Only some changes in docs.

**Pull request merged**